// ambient-cbg-ts/lib/esm/utils/Math.js
// Import what we need from Match class.
var PI = Math.PI, abs = Math.abs, sqrt = Math.sqrt, pow = Math.pow, random = Math.random, atan2 = Math.atan2;
var HALF_PI = 0.5 * PI;
var TAU = 2 * PI;
var TO_RAD = PI / 180;
var floor = function (n) { return n | 0; };
var rand = function (n) { return n * random(); };
var randIn = function (min, max) { return rand(max - min) + min; };
var randRange = function (n) { return n - rand(2 * n); };
var fadeIn = function (t, m) { return t / m; };
var fadeOut = function (t, m) { return (m - t) / m; };
var fadeInOut = function (t, m) {
    var hm = 0.5 * m;
    return abs(((t + hm) % m) - hm) / hm;
};
var dist = function (x1, y1, x2, y2) {
    return sqrt(pow(x2 - x1, 2) + pow(y2 - y1, 2));
};
var angle = function (x1, y1, x2, y2) {
    return atan2(y2 - y1, x2 - x1);
};
var lerp = function (n1, n2, speed) {
    return (1 - speed) * n1 + speed * n2;
};
export { HALF_PI, TAU, TO_RAD, floor, rand, randIn, randRange, fadeIn, fadeOut, fadeInOut, dist, angle, lerp, };
