// ambient-cbg-ts/lib/cjs/utils/Math.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.lerp = exports.angle = exports.dist = exports.fadeInOut = exports.fadeOut = exports.fadeIn = exports.randRange = exports.randIn = exports.rand = exports.floor = exports.TO_RAD = exports.TAU = exports.HALF_PI = void 0;
// Import what we need from Match class.
var PI = Math.PI, abs = Math.abs, sqrt = Math.sqrt, pow = Math.pow, random = Math.random, atan2 = Math.atan2;
var HALF_PI = 0.5 * PI;
exports.HALF_PI = HALF_PI;
var TAU = 2 * PI;
exports.TAU = TAU;
var TO_RAD = PI / 180;
exports.TO_RAD = TO_RAD;
var floor = function (n) { return n | 0; };
exports.floor = floor;
var rand = function (n) { return n * random(); };
exports.rand = rand;
var randIn = function (min, max) { return rand(max - min) + min; };
exports.randIn = randIn;
var randRange = function (n) { return n - rand(2 * n); };
exports.randRange = randRange;
var fadeIn = function (t, m) { return t / m; };
exports.fadeIn = fadeIn;
var fadeOut = function (t, m) { return (m - t) / m; };
exports.fadeOut = fadeOut;
var fadeInOut = function (t, m) {
    var hm = 0.5 * m;
    return abs(((t + hm) % m) - hm) / hm;
};
exports.fadeInOut = fadeInOut;
var dist = function (x1, y1, x2, y2) {
    return sqrt(pow(x2 - x1, 2) + pow(y2 - y1, 2));
};
exports.dist = dist;
var angle = function (x1, y1, x2, y2) {
    return atan2(y2 - y1, x2 - x1);
};
exports.angle = angle;
var lerp = function (n1, n2, speed) {
    return (1 - speed) * n1 + speed * n2;
};
exports.lerp = lerp;
